Create testing standard (#320)
UKHomeOffice/engineering-guidance-and-standards: docs/standards/testing.md

  - Testing
related:
  sections:
    - title: Related links
      items:
---

Testing code forms part of the core of developing code. Testing your changes is necessary to gain assurance that they can be deployed safely to production and that the service will still function as expected.

The intent of any test should be clear as they can provide a form of documentation to the overall solution, they should be relatively easy to understand with coverage tracked, although it must be stressed that coverage is not the be all end all.

A good test should be clear on the intent of the test, should only have one test case, and be readable and the test should pass everytime without changes to the underlying code.

Testing can come in different forms unit/integration/mutation/property based, but all must adhere to the requirements laid out as the following benefits can be achieved from it:

- Easy to connect failures in code from failed tests
- Security & confidence when continously integrating code
- Easy to understand intent and functionality of code
- Tests should be automated when possible
- Developers should be able to write and run tests and not silo'd to QAT

---

## Requirement(s)

> - [You MUST test early and often](#you-must-test-early-and-often)
> - [You MUST have a way of measuring the effectiveness of testing](#you-must-have-a-way-of-measuring-the-effectiveness-of-testing)
> - [It MUST be clear what is being tested and what the results are](#it-must-be-clear-what-is-being-tested-and-what-the-results-are)
> - [Tests MUST be repeatable](#tests-must-be-repeatable)
> - [Tests MUST be included for both the happy path and unhappy path](#tests-must-be-included-for-both-the-happy-path-and-unhappy-path)
> - [Tests that block deployments MUST complete in a reasonable time](#tests-that-block-deployments-must-complete-in-a-reasonable-time)
> - [You MUST use automated security testing](#you-must-use-automated-security-testing)
> - [You MUST make sure tests fail](#you-must-make-sure-tests-fail)

Author: ```suggestion
- [You MUST test early and often](#you-must-test-early-and-often)
- [You MUST automate tests]
- [Tests MUST be repeatable](#tests-must-be-repeatable)
- [You MUST have a way of measuring the effectiveness of testing](#you-must-have-a-way-of-measuring-the-effectiveness-of-testing)
- [It MUST be clear what is being tested and what the results are](#it-must-be-clear-what-is-being-tested-and-what-the-results-are)
- [Tests MUST think about the edge cases](#tests-must-be-included-for-both-the-happy-path-and-unhappy-path)
```

Add a pattern for what makes a test repeatable.
Call out security testing in automated testing.
Avoid false positives pattern.

### You MUST test early and often

The earlier testing happens in the development process, the less time is spent debugging, the better quality of the code through more tests, and the different paths are considered more. Using TDD can help here, as you can test the logic before writing it, which can help develop logic and sets the foundations of what you expect your code to do.

### You MUST have a way of measuring the effectiveness of testing

> You need to know ehether your test effictiveness has changed over the course of your developments and that you haven't had any regressions on your changes. Tracking this effectiveness over times helps to understand the landscape of your testing moving forward.

Author: ```suggestion
You need to know whether your test effectiveness has changed over the course of your developments and that you haven't had any regressions on your changes. Tracking this effectiveness over time helps to understand the landscape of your testing moving forward.
```

> One way of doing this is through monitoring test coverage, Test coverage helps a team understand the amount of code that is tested within a codebase. Setting a minimum threshold (such as 80%) ensures that any new code is tested. Another example is mutation testing.

Author: ```suggestion
One way of doing this is through monitoring test coverage, test coverage helps a team understand the amount of code that is tested within a codebase. Setting a minimum threshold (such as 80%) ensures that any new code is tested. Another example is mutation testing.
```

### You MUST use automated security testing

Automated security testing can help understand where your code may have any security issues. These tests should be be ran as part of your pipelines to help identify issues and block the deployment of insecure code.

Automated security testing, good peer reviews and good developer knowledge around keeping applications secure help to make sure an application stays secure.

A good example of this is using Sonarqube as part of your pipelines.

### You MUST make sure tests fail

Making sure tests fail helps to verify that your tests are doing the correct thing and that they would fail if there is an issue.

One way of achieving this is using red/green testing, when using TDD you get this for free, as you write the test first, have it fail (Red), then write the code to make the test pass (Green).

Another way of making sure tests fail is using mutation testing. Mutation testing creates small varitions of source code and subjects them to the same tests and dataset to detect any errors in the code.

### You MUST use automated accessibility testing

> Accessibility testing can help understand where any accessibility issuses may be in your code. If there is a failure, it shouldblock your code from deploying so you have opportunity to fix any issues first.

Reviewer: ```suggestion
Accessibility testing can help understand where any accessibility issues may be in your code. If there is a failure, it should block your code from deploying so you have opportunity to fix any issues first.
```

> Accessibility testing can help understand where any accessibility issuses may be in your code. If there is a failure, it shouldblock your code from deploying so you have opportunity to fix any issues first.

Reviewer: Should this completely block deployments?

---
layout: standard
order: 1
title: Developer Testing
date: 2023-10-26
id: SEGAS-TOBEFILLED
tags:
  - Source management
  - Secure development
  - Ways of working
  - Testing
related:
  sections:
    - title: Related links
      items:
---

> Testing code forms part of the core of developing code. Testing should provide confidence to stakeholders and be proportionate to risk.

Reviewer 2: Confidence of what?

If the team is responsible for the production service, is this about external stakeholders or is it about the team themselves knowing whether a change is safe?

Author: @daniel-ac-martin this was for example, to show a PO that changes are not going to break things, as we have tests in place to help cover it.

Reviewer 2: With regards to 'confidence', I think it's about gaining assurance that the software/service still correctly performs the functions that it needs to. i.e. That it is 'safe' to deploy the change.

With regards to stakeholders, I think I would avoid the term, as it can be misinterpreted. Basically, the key thing is that we need to aim for this sort of assurance to sit with the product team. I don't even think that a PO should need to worry about this if he/she has faith the the developers. The assurance is really for the developers themselves, as they are the ones that are responsible for the production service.

I think I'd go for _something_ like:
> Testing your changes is necessary to gain assurance that they can be deployed safely to production and that the service will still function as expected.

> The intent of any test should be clear as they can provide as a form of documentation to the overall solution, they should be relatively easy to understand with coverage tracked, though this must be stressed coverage is not the be all end all.

Author: Should include that it's a programmatic tests.

> The intent of any test should be clear as they can provide as a form of documentation to the overall solution, they should be relatively easy to understand with coverage tracked, though this must be stressed coverage is not the be all end all.

Reviewer 2: ```suggestion
The intent of any test should be clear as they can provide a form of documentation to the overall solution, they should be relatively easy to understand with coverage tracked, although it must be stressed that coverage is not the be all end all.
```

A good test should be clear on the intent of the test, should only have one test case, and be readable and the test should pass everytime without changes to the underlying code.

Testing can come in different forms unit/integration/mutation/property based, but all must adhere to the requirements laid out as the following benefits can be achieved from it:

- Easy to connect failures in code from failed tests
- Security & confidence when continously integrating code
- Easy to understand intent and functionality of code

---

## Requirement(s)

- [You MUST track test coverage](#you-must-track-test-coverage)
- [It MUST be clear when and why tests fail](#it-must-be-clear-when-and-why-tests-fail)
- [You MUST test early and often](#you-must-test-early-and-often)
- [Tests MUST be repeatable](#tests-must-be-repeatable)
- [Tests MUST be included for both the happy path and unhappy path](#tests-must-be-included-for-both-the-happy-path-and-unhappy-path)
- [Tests MUST be fast](#tests-must-be-fast)
- [You MUST test for security issues as part of your pipelines](#you-must-test-for-security-issues-as-part-of-your-pipelines)

> ### You MUST track test coverage
> 
> Test coverage helps a team understand the amount of code that is tested within a codebase. Setting a minimum threshold (such as 80%) ensures that any new code is tested. Coverage is not always accurate, as if a test has an always truthy assertion then coverage can over estimate. However, it is a good indication overall.

Author: ```suggestion
### You MUST have a way of measuring the effectiveness of testing 

Test coverage helps a team understand the amount of code that is tested within a codebase. Setting a minimum threshold (such as 80%) ensures that any new code is tested. Coverage is not always accurate, as if a test has an always truthy assertion then coverage can over estimate. However, it is a good indication overall.
```

Author: Test early and often

Author: Add something around track past data

Author: Other examples e.g. mutation testing.

### It MUST be clear when and why tests fail.

> Understanding why tests are failing and been transparent about why (such as a report, or CI pipeline) can help the whole team see the value, and understand where failures are occuring. It also builds confidence within the team and business when tests wither catch something, or that they are running regularily. You shouldn't need to check manually whether a test has passed or failed.

Author: When a pipeline fails, should be able to see what has failed,

Author: ```suggestion
### It MUST be clear what test results are.

Understanding why tests are failing and been transparent about why (such as a report, or CI pipeline) can help the whole team see the value, and understand where failures are occurring. It also builds confidence within the team and business when tests wither catch something, or that they are running regularly. You shouldn't need to check manually whether a test has passed or failed.
```

### You MUST test early and often

The earlier testing happens in the development process, the less time is spent debugging, the better quality of the code through more tests, and the different paths are considered more. Using TDD can help here, as you can test the logic before writing it, which can help develop logic and sets the foundations of what you expect your code to do.

### Tests MUST be repeatable

The ability to repeat the same tests, gives confidence that if there is a failure, something has changed to affect that failure. The values shouldn't change based on enviroments and unit tests should not depend on other external factors to run.

### Tests MUST be included for both the happy path and unhappy path

> Testing for all eventualities helps to write code that can handle errors in a better way. This should include things like testing for illegal arguements, or no arguements to be passed.

Reviewer 2: I think that there is a problem here in that there can be many unhappy paths.

I would go for something like this:
- MUST test the happy path
- MUST test security features (e.g. ensure data can't be seen when not logged in)
- SHOULD test common unhappy paths

As with all of this there's also the question of what is proportionate and at what time in a project's life-cycle.

Author: Why is it only a SHOULD that we test unhappy paths? Arguably, these would be more important?

Reviewer 2: We might be running into semantics here, and what counts as an unhappy path. The main thing I'm worried about is fractal complexity as there can be many more unhappy paths than happy ones. (Consider form validation, there might be one valid way through but each field might have many different ways to be invalid.)

I do think it's important to test the more common unhappy paths.

But if I had a service that is very early in its life-cycle (and subject to rapid change), I would be more worried about maintaining the happy path than whether things go wrong in the correct way. (The big exception being security features.)

Author: I'm going to take these suggestions to the next SD guild meeting and discuss them further and get the opinions of the group.

> ### Tests MUST be fast
> 
> A developer should be able to run tests quickly to help understand the inpact of change. The longer it takes for a suite of tests to run, the less likely a developer is going to run the entire suite, which might cause issues later on in the development cycle.

Author: Main reason is to build confidence.
```suggestion
### Tests MUST complete in a reasonable time

A developer should be able to run tests quickly to help understand the impact of change. The longer it takes for a suite of tests to run, the less likely a developer is going to run the entire suite, which might cause issues later on in the development cycle.
```

Reviewer 2: I think this statement is a bit too strong.

We need automated tests that run as a part of the overall pipeline, of getting a change out to production, to be fast. (And what counts as fast probably depends on the complexity of the project and the appetite for risk.)

That said, I think there's probably still a case for longer running tests that do NOT block deployments, but instead alert the team whenever they fail. i.e. Changes don't wait for them, but if they fail that causes an incident that must be addressed by the team.

Soak tests would fall into this category, for example.

Author: @daniel-ac-martin Yeah, we spoke about this, and said this applies more to unit/integration tests, as there are other types of testing that can take longer. I wonder if we can word it differently?

### You MUST test for security issues as part of your pipelines

> Testing security issues in your pipeline is vital to help make sure your application is secure and doesn't have any vulnerabilities.

Author: Code smells etc.

> Testing security issues in your pipeline is vital to help make sure your application is secure and doesn't have any vulnerabilities.

Reviewer 2: I think we should be making use of automated security testing, but the above statement is untrue. - It is possible to write secure software without such testing.

I think it's also important that we avoid teams having a false sense of security about such testing. - They can only catch some of the obvious things. We still need good peer review, and we still need developers to think about security and what could go wrong.

Author: @daniel-ac-martin agreed, I will look at rephrasing this

> 

Author: Add something about making sure tests can fail

Author: Mutation tests can help with this.

Reviewer 2: That's normally called red/green testing I think.
Obviously it comes for free with TDD.

> ---

Reviewer 3: Do we need something like (but a lot more eloquent and succint)

## Tests should be considered failing or in need of review if they demonstrate a significant change in execution time when a code change is made

Maybe `Tests MUST be treated as failed on Unintended Performance Deviations`

Reviewer 2: This would be nice, but can we currently do it across the board?

If not, I don't think it's right to have it as a MUST.

Reviewer 2: Hmm. I think this is partially related to the guidance vs standards question.

I've already got a lot of the guidance that I think I need for my area, so I'm mainly looking for a set of formal standards to assess projects against.

But there is a strong case to be made for trying to bring them together. I think that would be a question of how we present that.

As for more aspirational ideas ("Wouldn't it be nice if") that potentially could like alongside guidance, alternatively it could be something that we track in issues against this repo.

---
layout: standard
order: 1
title: Developer Testing
date: 2023-10-26
id: SEGAS-TOBEFILLED
tags:
  - Source management
  - Secure development
  - Ways of working
  - Testing
related:
  sections:
    - title: Related links
      items:
---

Testing code forms part of the core of developing code. Testing your changes is necessary to gain assurance that they can be deployed safely to production and that the service will still function as expected.

> The intent of any test should be clear as they can provide a form of documentation to the overall solution, they should be relatively easy to understand with coverage tracked, although it must be stressed that coverage is not the be all end all.

Reviewer 4: This sentence is clunky and hard to read.  I think the grammar needs fixing.  Suggestion:

The intent of any test should be clear as they can provide a form of documentation to the overall solution.
The tests should be relatively easy to understand with a track of the overall testing coverage kept.  It is good to keep in mind that test coverage of a code base is not a definitive marker of quality.

Author: Looks good to me, I'll update

> A good test should be clear on the intent of the test, should only have one test case, and be readable and the test should pass everytime without changes to the underlying code.

Reviewer 4: Correcting grammar suggestion:

A good test should be clear on the intent of the test, have one test case, be readable and should pass every time without changes to the underlying code.

> Testing can come in different forms unit/integration/mutation/property based, but all must adhere to the requirements laid out as the following benefits can be achieved from it:

Reviewer 4: Correcting grammar suggestion:

Testing can come in different forms, unit/integration/mutation/property based, but all must adhere to the requirements as the following benefits can be achieved:

> - Easy to connect failures in code from failed tests

Reviewer 4: I think this is meant to be correct:

Easy to correct failures in code from failed tests.

- Security & confidence when continously integrating code
- Easy to understand intent and functionality of code

> - Tests should be automated when possible

Reviewer 4: Not sure this should be here as a benefit, it is a requirement.

> - Developers should be able to write and run tests and not silo'd to QAT

Reviewer 4: Not sure this is a benefit either as it reads like a requirement.  Maybe it could be re-written as:

- Tests are owned by the code owners and not within a silo external to the team.

Happy to be challenged a bit on this one though.  It could probably be a bit more clear.

Reviewer 4: Agreed looks good.

Reviewer 5: This touches on code ownership, but is code always owned by the same team or by a team? - e.g. more open source projects for instance may not always be owned by a team.

Also, shouldn't this be generally left to a team or project's Ways of Working - i.e who writes what test - within a team or outside? - as I agree that the above point around this is not a benefit of having tests.

_Not sure this should be part of this standard, and instead should be covered by any other standard that caters more of general code ownership - or perhaps a standard/principle that generally covers engineering team dynamics - if there is one._

---

## Requirement(s)

- [You MUST test early and often](#you-must-test-early-and-often)
- [You MUST automate tests](#you-must-automate-tests)
- [You MUST make tests repeatable](#you-must-make-tests-repeatable)
- [You MUST have a way of measuring the effectiveness of testing](#you-must-have-a-way-of-measuring-the-effectiveness-of-testing)

> - [You MUST make it clear what is being tested and what the results are](#you-must-make-it-clear-what-is-being-tested-and-what-the-results-are)

Author: ```suggestion
- [Tests MUST have a purpose and explicit result](#tests-must-have-a-purpose-and-explicit-result)
```

- [You MUST think about the edge cases](#you-must-think-about-the-edge-cases)

### You MUST test early and often

The earlier testing happens in the development process, the less time is spent debugging, the better quality of the code through more tests, and the different paths are considered more. Using TDD can help here, as you can test the logic before writing it, which can help develop logic and sets the foundations of what you expect your code to do.

### You MUST automate tests

> Automating tests enabled the tests to be ran often, and block pipelines that fail testing. It also means code is tested often, making sure that everything is working as expected.

Reviewer 6: ```suggestion
Automating tests enables the tests to be run consistently and often. Automated tests can be run as part of a build pipeline, and block the pipeline if they fail. It also means code is tested often, making sure that everything is working as expected.
```

### You MUST make tests repeatable

The ability to repeat the same tests, gives confidence that if there is a failure, something has changed to affect that failure. The values shouldn't change based on enviroments and unit tests should not depend on other external factors to run.

### You MUST have a way of measuring the effectiveness of testing

You need to know whether your test effectiveness has changed over the course of your development and that you haven't had any regressions on your changes. Tracking this effectiveness over time helps to understand the landscape of your testing moving forward.

One way of doing this is through monitoring test coverage, Test coverage helps a team understand the amount of code that is tested within a codebase. Setting a minimum threshold (such as 80%) ensures that any new code is tested. Another example is mutation testing.

### Tests MUST have a purpose and explicit result

Tests must indicate why they are failing, showing expected and actual responses and the name of the test.

An example is through making sure pipelines fail when tests fail, and you're able to see more information about why.

> 

Reviewer 6: ```suggestion
### Tests MUST have a purpose and explicit result

Tests must indicate why they are failing, showing expected and actual responses and the name of the test.

An example is through making sure pipelines fail when tests fail, and you're able to see more information about why.

A test failing due to a regression often exposes an engineer to functionality they are not familiar with. The information logged along with the test implementation should provide enough context that they can understand what is being tested and what has gone wrong, and be able to reason about how they can resolve it.

> ### You MUST make it clear what is being tested and what the results are
> 
> Understanding why tests are failing and been transparent about why can help the whole team see the value, and understand where failures are occurring. You shouldn't need to chesck manually that tests have failed.
> 
> An example is through making sure piplelines fail when tests fail, and you're able to see more information about why.

Author: ```suggestion
### Tests MUST have a purpose and explicit result

Tests must indicate why they are failing, showing expected and actual responses and the method under test.

An example is through making sure pipelines fail when tests fail, and you're able to see more information about why.
```

---
layout: standard
order: 1
title: Developer Testing
date: 2023-10-26
id: SEGAS-TOBEFILLED
tags:
  - Source management
  - Secure development
  - Ways of working
  - Testing
related:
  sections:
    - title: Related links
      items:
---

Testing code forms part of the core of developing code. Testing your changes is necessary to gain assurance that they can be deployed safely to production and that the service will still function as expected.

The intent of any test should be clear as they can provide a form of documentation to the overall solution.

> The tests should be relatively easy to understand with a track of the overall testing coverage kept. It is good to keep in mind that test coverage of a code base is not a definitive marker of quality.

Reviewer 5: I'll reword this "It is good to keep in mind that test coverage of a code base is not a definitive marker of quality."

To something like: "Strong test coverage of a code base contributes towards overall quality of application code, although it is not necessarily the only definitive marker of quality code depending on your application context."

_i.e We appreciate that it's not all about test coverage depending on your use case, but we don't want to seemingly discourage people from making an effort towards having good coverage where possible._

Reviewer 4: Good points, agreed.

Author: Agreed && updated

A good test should be clear on the intent of the test, have one test case, be readable and should pass every time without changes to the underlying code.

Testing can come in different forms, unit/integration/mutation/property based, but all must adhere to the requirements as the following benefits can be achieved:

- Easy to correct failures in code from failed tests.
- Security & confidence when continously integrating code
- Easy to understand intent and functionality of code
- Developers should be able to write and run tests and not silo'd to QAT

---

## Requirement(s)

- [You MUST test early and often](#you-must-test-early-and-often)
- [You MUST automate tests](#you-must-automate-tests)
- [You MUST make tests repeatable](#you-must-make-tests-repeatable)
- [You MUST have a way of measuring the effectiveness of testing](#you-must-have-a-way-of-measuring-the-effectiveness-of-testing)
- [Tests MUST have a purpose and explicit result](#tests-must-have-a-purpose-and-explicit-result)
- [You MUST think about the edge cases](#you-must-think-about-the-edge-cases)

### You MUST test early and often

> The earlier testing happens in the development process, the less time is spent debugging, the better quality of the code through more tests, and the different paths are considered more. Using TDD can help here, as you can test the logic before writing it, which can help develop logic and sets the foundations of what you expect your code to do.

Reviewer 5: Consider restructuring to something like this:

```suggestion
Writing tests early and often in a software development process tends to reduce the amount of time spent debugging. Good tests written early generally helps improve the overall assurance and design quality of your application. 
Test Driven Development (TDD) is a valuable development approach that can help towards achieving this. This development approach encourages you to think ahead on how your requirements can be met by thinking on how your solution can be tested first at unit level. Once a test is written first albeit failing due to non-existing solution, you can implement your solution to pass the test, and then refactor. "Testing early and often" in this case leads to "refactoring early and often", which continuously improves the quality of your application code and design.```

### You MUST automate tests

Automating tests enables the tests to be run consistently and often. Automated tests can be run as part of a build pipeline, and block the pipeline if they fail. It also means code is tested often, making sure that everything is working as expected.

### You MUST make tests repeatable

> The ability to repeat the same tests, gives confidence that if there is a failure, something has changed to affect that failure. The values shouldn't change based on enviroments and unit tests should not depend on other external factors to run.

Reviewer 5: ```suggestion
The ability to repeat the same tests gives confidence that if there is a failure, something has changed to affect that failure. The values that a test rely on shouldn't change based on environments, and tests such as unit tests should not depend on other external factors and dependencies (such as third party API, etc) to run.
```

### You MUST have a way of measuring the effectiveness of testing

You need to know whether your test effectiveness has changed over the course of your development and that you haven't had any regressions on your changes. Tracking this effectiveness over time helps to understand the landscape of your testing moving forward.

One way of doing this is through monitoring test coverage, Test coverage helps a team understand the amount of code that is tested within a codebase. Setting a minimum threshold (such as 80%) ensures that any new code is tested. Another example is mutation testing.

### Tests MUST have a purpose and explicit result

Tests must indicate why they are failing, showing expected and actual responses and the name of the test.

An example is through making sure pipelines fail when tests fail, and you're able to see more information about why.

> A test failing due to a regression often exposes an engineer to functionality they are not familiar with. The information logged along with the test implementation should provide enough context that they can understand what is being tested and what has gone wrong, and be able to reason about how they can resolve it.

Reviewer 5: ```suggestion
A test failing due to a regression often implies that a functionality has been added that contradict existing (and previous expectations), this should be reviewed to understand why this happened and what needs to be done to fix the regression issue. Although the area of concern may be unfamiliar, the information logged along with the test implementation should provide enough context to help with this review - in understanding what is being tested, what has gone wrong, and hence be able to reason about how the issue can be resolved.
```

Reviewer 6: My intent with this paragraph was to explain why it is important that "Tests MUST have a purpose and explicit result" which is to give context to an engineer interacting with the test and the behaviour it is verifying. It is targetted at the author of the test, so adding notes about how the engineer who sees the failing test should respond confuses things.

### You MUST think about the edge cases

> Testing for all eventualities helps to write code that can handle errors in a better way. This should include things like testing for illegal arguements, or no arguements to be passed.

Reviewer 5: ```suggestion
Testing for all eventualities helps to write code that can handle errors and exceptions. This should include things like testing for illegal or no arguments, dependencies' edge cases (which can often be simulated through mocking dependencies as appropriate).
```
